controller/sharding/cache.go

```diff
@@ -0,0 +1,169 @@
+package sharding
+
+import (
+	"sync"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/db"
+	log "github.com/sirupsen/logrus"
+)
+
+type ClusterSharding interface {
+	Init(clusters *v1alpha1.ClusterList)
+	Add(c *v1alpha1.Cluster)
+	Delete(clusterServer string)
+	Update(c *v1alpha1.Cluster)
+	IsManagedCluster(c *v1alpha1.Cluster) bool
+	GetDistribution() map[string]int
+}
+
+type clusterSharding struct {
+	shard           int
+	replicas        int
+	shards          map[string]int
+	clusters        map[string]*v1alpha1.Cluster
+	lock            sync.RWMutex
+	getClusterShard DistributionFunction
+}
+
+func NewClusterSharding(db db.ArgoDB, shard, replicas int, shardingAlgorithm string) ClusterSharding {
+
+	log.Infof("Processing clusters from shard %d", shard)
+	log.Infof("Using filter function:  %s", shardingAlgorithm)
+	clusterSharding := &clusterSharding{
+		shard:    shard,
+		replicas: replicas,
+		shards:   make(map[string]int),
+		clusters: make(map[string]*v1alpha1.Cluster),
+	}
+
+	distributionFunction := NoShardingDistributionFunction(0)
```

Comment:
Should we check for `shardingAlgorithm  = NoShardingAlgorithm` here as well as we are always going to return the same shard?

Reply:
I will go and change the function signature to have no argument then